indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveUpgradedToSegmentsIdsAction.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.actions;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.metadata.SegmentUpgradeInfo;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Task action to retrieve the segment upgrade infos of a given set of used or unused upgraded segment ids.
+ */
+public class RetrieveUpgradedToSegmentsIdsAction implements TaskAction<List<SegmentUpgradeInfo>>
+{
+  private final String dataSource;
+  private final Set<String> upgradedSegmentIds;
+
+  @JsonCreator
+  public RetrieveUpgradedToSegmentsIdsAction(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("upgradedSegmentIds") Set<String> upgradedSegmentIds
+  )
+  {
+    this.dataSource = dataSource;
+    this.upgradedSegmentIds = upgradedSegmentIds;
+  }
+
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  public Set<String> getUpgradedSegmentIds()
+  {
+    return upgradedSegmentIds;
+  }
+
+  @Override
+  public TypeReference<List<SegmentUpgradeInfo>> getReturnTypeReference()
+  {
+    return new TypeReference<List<SegmentUpgradeInfo>>()
+    {
+    };
+  }
+
+  @Override
+  public List<SegmentUpgradeInfo> perform(Task task, TaskActionToolbox toolbox)
+  {
+    return toolbox.getIndexerMetadataStorageCoordinator()
```

Comment:
Along with checking for children, we should also check if the root segment IDs are themselves present in the database or not. At the `IndexerMetadataStorageCoordinator` layer, they should be two separate methods but the task action can return them in a single map.


```java
{
   "a0" : ["a1"],          // only reference is a child 
   "b0" : ["b0", "b1"],    // two references - child and self
   "c0" : null             // no reference, segment files can be cleaned up
}
```

Reply:
For this purpose, you could reuse the `retrieveSegmentsById` method. Alternatively, you could just fire the action `RetrieveSegmentsByIdAction` from the `kill` task and build the full picture on the task itself.

Reply:
>  Could you please suggest a way to test these changes so that such errors do not creep in?

Write out a unit test that ensures that the required task actions are fired.

>  I was thinking we could create a separate method nukeSegmentsAndDetermineSubsetToKill

Do you mean you would like to test this method by directly invoking it from the unit test? If yes, then I would advise against that.

Reply:
> Write out a unit test that ensures that the required task actions are fired.

This would have still missed the scenario of parents being present.

Reply:
You should write a test where a child is being killed but parent still exists in the metadata store.
The test should then ensure that even in this case, the segment files are not deleted from deep store.
Let me know if this helps.

Reply:
I feel that is a complicated way to test something that could be done with a unit test by separating the nuke and segments to kill compuation to a package private method. Could you please share why you advise against it?
The test that you suggest seems better suited for an integration test.

Reply:
> The test that you suggest seems better suited for an integration test.

Is it not possible to do this assertion in a unit test?

Reply:
> I feel that is a complicated way to test something that could be done with a unit test by separating the nuke and segments to kill compuation to a package private method. Could you please share why you advise against it?

The job of a unit test or any test is to verify the overall contract of a module or class in different scenarios.
We should be able to control the behaviour of the module using only the inputs we give it. The rest of it should be a blackbox for the test.
The test is not meant to verify the behaviour of every single internal method.

You could even try test driven design for this purpose. Write out the tests first and then the internal implementation. Otherwise, we will always end up exposing internal details.

The case at hand is a fairly simple and shouldn't even require so much thought.